### ISSUE TITLE
More of a feature request than an issue

### ISSUE THREAD

**Issue by mbechc** [bartbutenaers/node-red-contrib-button-events#12]
Sorry, didn't know how to else get in contact and suggest a feature.

If one has a button up and button down switch for controlling light [like this](https://www.smartkontakten.dk/vare/zigbee-tryk-med-dobbelt-taend-sluk-til-lk-fuga-hv/?attribute_farve=Hvid&gad_source=1&gclid=Cj0KCQjwqP2pBhDMARIsAJQ0Czom-saDqNxFMQFHHOflxbEB9ps1lSgCgh-G3_UE3Aw2GGPvdXmpskkaAtlBEALw_wcB)

The natural thing would be to turn up brightness to 100% when pressing 1, and to 0% brightness when pushing 0

The fantastic thing would be to start emmit a "signal" out of the button node, starting at fx 500ms, 250ms later a new "signal" - this would allow to step up (or down) lights by fx 10% for every up/down emitted while keeping the button pressed.

This button sends a Pressed signal until button is released, then a release signal. if the button is pressed and released fast it must mean a press - that is covered in your node. I would prefer that Long press is replaced by continuously emitted  steps, in stead of a single Long.

Short press on 1 - Turn on light
Continous press on 0 - step down, step down, step down

Or another scenario

Lights are off 
A continous press on button 1 - step up, step up step up, until button released

I know, your node doesn't care about button 1 vs button 0, so I would have to program the node to either emmit up or down, or simply have the node emit - step if button is continously pressed (ie button release event is not given within time fx 500ms)

Hope you will extend your node, I simply cannot make this using regular nodes/flows

**Comment by geoffreydemaagd** [comment]
> Sorry, didn't know how to else get in contact and suggest a feature.
> 
> If one has a button up and button down switch for controlling light [like this](https://www.smartkontakten.dk/vare/zigbee-tryk-med-dobbelt-taend-sluk-til-lk-fuga-hv/?attribute_farve=Hvid&gad_source=1&gclid=Cj0KCQjwqP2pBhDMARIsAJQ0Czom-saDqNxFMQFHHOflxbEB9ps1lSgCgh-G3_UE3Aw2GGPvdXmpskkaAtlBEALw_wcB)
> 
> The natural thing would be to turn up brightness to 100% when pressing 1, and to 0% brightness when pushing 0
> 
> The fantastic thing would be to start emmit a "signal" out of the button node, starting at fx 500ms, 250ms later a new "signal" - this would allow to step up (or down) lights by fx 10% for every up/down emitted while keeping the button pressed.
> 
> This button sends a Pressed signal until button is released, then a release signal. if the button is pressed and released fast it must mean a press - that is covered in your node. I would prefer that Long press is replaced by continuously emitted steps, in stead of a single Long.
> 
> Short press on 1 - Turn on light Continous press on 0 - step down, step down, step down
> 
> Or another scenario
> 
> Lights are off A continous press on button 1 - step up, step up step up, until button released
> 
> I know, your node doesn't care about button 1 vs button 0, so I would have to program the node to either emmit up or down, or simply have the node emit - step if button is continously pressed (ie button release event is not given within time fx 500ms)
> 
> Hope you will extend your node, I simply cannot make this using regular nodes/flows

Why don't you use the node https://flows.nodered.org/node/node-red-contrib-time-based-dimmer to fulfill this requirement?
I've just the same setup as you described and using both nodes to achieve this behaviour.

**Comment by bartbutenaers** [comment]
@mbechc,
Sorry for the delay!  Kind of busy lately...

I like the proposal from @geoffreydemaagd.  Thanks Geoffrey for sharing your ideas!!!!!!!!!!!!!
Bart

**Comment by mbechc** [comment]
Yes, thanks both for chiming in. The reason why "time based dimmer" doesn't work is that it lacks thing number one, 

**It doesn't do this - which I feel is essential for a lightbulb switch**
1 FAST push (on button marked 1) turn to 100%
1 FAST push (on button marked 0) turn to 0%

**This is what "time based dimmer does"**
Long push (on button marked 1) steps up and vice versa, down

That is why I asked you Bart, because your code knows about short vs long press, ideally your code would emit, step up or step down continuously if button is pressed long. I have tested your code, to only emit a single Long

A merge between your code - long press, and "Time based dimmer" would be perfect to control light bulbs, with dimming function 

Thanks Guys 

/Morten

**Comment by bartbutenaers** [comment]
@mbechc,

Yes I completely understand your proposal.  I have a physical dimmer from Niko at my home which behaves exactly like in your description.

However I won't merge the time based dimmer code in my node for a couple of reasons:
+ Severe lack of free time
+ I prefer to keep responsibilities separated.  This node detects types of button events.
+ My node is build on top of another library, so I have no control about most of the code.  If I had to start all over again, I would most probably have done it completely otherwise.

@geoffreydemaagd: do you know a way how to solve this?  A simplified example flow would be very welcome...

**Comment by geoffreydemaagd** [comment]
For all my pushbuttons I use following setup

![image](https://github.com/bartbutenaers/node-red-contrib-button-events/assets/59262677/6222a936-f256-42d3-ad86-31d114856db8)


[{"id":"2eba858d508e82e0","type":"button-events","z":"d73909db3492c62d","name":"WC Spot","outputs":3,"inputField":"payload","inputFieldType":"msg","outputField":"payload","outputFieldType":"msg","downValue":"1","downValueType":"num","upValue":"0","upValueType":"num","buttonEventsConfig":"d1c95541c1d66a92","clickedInterval":null,"pressedInterval":null,"debounceInterval":null,"events":[{"type":"clicked"},{"type":"pressed"},{"type":"released"}],"x":360,"y":200,"wires":[["d2b1c198357ffa11"],["d59ffe978cb765d8","9e32ee938a358f5f"],["d59ffe978cb765d8"]]},{"id":"d2b1c198357ffa11","type":"function","z":"d73909db3492c62d","name":"ON/OFF","func":"var spot_wc_0 = global.get('spot_wc_0_status');\nif (spot_wc_0 == undefined) {\n  spot_wc_0 = false;\n}\n\nif (msg.payload == 'clicked') {\n\n  msg.payload = !spot_wc_0;\n}\n\n\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":560,"y":180,"wires":[["99bde571b61a774c"]]},{"id":"99bde571b61a774c","type":"link out","z":"d73909db3492c62d","name":"Pushbutton_Toggle_ws_spot_0_OUT","links":["a6b0d897.4d0528"],"x":655,"y":180,"wires":[]},{"id":"d59ffe978cb765d8","type":"function","z":"d73909db3492c62d","name":"Dimming","func":"var spot_wc_0_updown = global.get('spot_wc_0_updown') || 'DOWN';\n\nif (spot_wc_0_updown === 'UP' && msg.payload === 'pressed') {\n    msg.payload = 'brightness';\n}\n\nelse if (spot_wc_0_updown === 'UP' && msg.payload === 'released') {\n    msg.payload = 'brightness_stop';\n    spot_wc_0_updown = 'DOWN';\n    global.set('spot_wc_0_updown', spot_wc_0_updown);\n}\n\nelse if (spot_wc_0_updown === 'DOWN' && msg.payload === 'pressed') {\n    msg.payload = 'brightness';\n\n}\n\nelse if (spot_wc_0_updown === 'DOWN' && msg.payload === 'released') {\n    msg.payload = 'brightness_stop';\n    spot_wc_0_updown = 'UP';\n    global.set('spot_wc_0_updown', spot_wc_0_updown);\n}\n\nnode.status(msg.payload);\n\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":560,"y":220,"wires":[["d3e01cfc0ea7180d"]]},{"id":"9e32ee938a358f5f","type":"function","z":"d73909db3492c62d","name":"Inject actual","func":"msg.payload = global.get('spot_wc_0');\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":570,"y":260,"wires":[["d3e01cfc0ea7180d"]]},{"id":"d3e01cfc0ea7180d","type":"one-button-dimmer","z":"d73909db3492c62d","name":"","interval":"20","step":"1","minValue":"5","maxValue":"255","startCommand":"brightness","stopCommand":"brightness_stop","x":740,"y":220,"wires":[["84e84ab664a2f2f8"]]},{"id":"84e84ab664a2f2f8","type":"function","z":"d73909db3492c62d","name":"Global","func":"global.set('spot_wc_0', parseInt(msg.payload));\nmsg.payload= parseInt(msg.payload);\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","x":930,"y":220,"wires":[["e4dbb34b1e966a2e"]]},{"id":"e4dbb34b1e966a2e","type":"link out","z":"d73909db3492c62d","name":"Pushbutton_Dimming_spot_wc_0_OUT","links":["5c4c0482.a3175c"],"x":1135,"y":220,"wires":[]},{"id":"d1c95541c1d66a92","type":"button-events-config","name":"config for 2d9ff7fc.daca08","clickedInterval":"1","pressedInterval":"1500","debounceInterval":"0"}]

Maybe you can use this or better, improve this flow?

**Comment by mbechc** [comment]
I gave it a try and did it using standard nodes - except the Timed Node - that counts the number of messages within a timeframe, its not exactly elegant yet

Button up and Button down, simply dims to either 0 or 100, but any of the two buttons can be double clicked (im counting the number of clicks, using Timed node) if double clicked, the buttons alter mode to Dimmable up and down, after 10 seconds, dimmable mode is exited, and buttons reverts back to simple 0 or 100 brightness.

<img width="761" alt="Screenshot 2023-11-12 at 15 43 48" src="https://github.com/bartbutenaers/node-red-contrib-button-events/assets/17846557/02addd1f-480d-4778-b47f-dd5f9d7687c3">



[
    {
        "id": "7bcf78a84f27cb72",
        "type": "tab",
        "label": "SW01",
        "disabled": false,
        "info": "",
        "env": []
    },
    {
        "id": "8e67781937d9e961",
        "type": "trigger",
        "z": "7bcf78a84f27cb72",
        "name": "",
        "op1": "25",
        "op2": "10",
        "op1type": "num",
        "op2type": "str",
        "duration": "-500",
        "extend": false,
        "overrideDelay": false,
        "units": "ms",
        "reset": "stop",
        "bytopic": "all",
        "topic": "topic",
        "outputs": 1,
        "x": 1020,
        "y": 60,
        "wires": [
            [
                "2097d594082c9566"
            ]
        ]
    },
    {
        "id": "2fb5329c471daf57",
        "type": "timed-counter",
        "z": "7bcf78a84f27cb72",
        "name": "detectDouble",
        "timelimit": "400",
        "timeunit": 1,
        "withhold": true,
        "fixedtimeout": true,
        "pertopic": true,
        "countlimit": "",
        "x": 560,
        "y": 240,
        "wires": [
            [
                "457d6fef26fcf7fb"
            ]
        ]
    },
    {
        "id": "457d6fef26fcf7fb",
        "type": "function",
        "z": "7bcf78a84f27cb72",
        "name": "Single or Doubleclick",
        "func": "if (msg.count==4) {\n    flow.set(\"dimmerMode\", 1)\n    msg = null;\n    }\n// Double press detected, entering dimmerMode\n\nelse if (msg.payload.action==\"release_1\") {msg.payload = 255}\n// On button pressed once\nelse if (msg.payload.action == \"release_2\") { msg.payload = 0 }\n// Off button pressed once\nreturn msg;",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 780,
        "y": 240,
        "wires": [
            [
                "9d09b86f2050d21b",
                "6ce773ae92d191c0"
            ]
        ]
    },
    {
        "id": "596121eb7d0e536d",
        "type": "switch",
        "z": "7bcf78a84f27cb72",
        "name": "dimmerMode",
        "property": "dimmerMode",
        "propertyType": "flow",
        "rules": [
            {
                "t": "eq",
                "v": "1",
                "vt": "num"
            },
            {
                "t": "else"
            }
        ],
        "checkall": "true",
        "repair": false,
        "outputs": 2,
        "x": 340,
        "y": 160,
        "wires": [
            [
                "1c7df14b7dcc4ddd"
            ],
            [
                "2fb5329c471daf57"
            ]
        ]
    },
    {
        "id": "9d09b86f2050d21b",
        "type": "trigger",
        "z": "7bcf78a84f27cb72",
        "name": "exit dimmerMode",
        "op1": "",
        "op2": "",
        "op1type": "nul",
        "op2type": "date",
        "duration": "20",
        "extend": false,
        "overrideDelay": false,
        "units": "s",
        "reset": "",
        "bytopic": "all",
        "topic": "topic",
        "outputs": 1,
        "x": 1010,
        "y": 200,
        "wires": [
            [
                "25038cc00759fcae"
            ]
        ]
    },
    {
        "id": "265ec2fc93d12ee3",
        "type": "link in",
        "z": "7bcf78a84f27cb72",
        "name": "link in 4",
        "links": [
            "18a33da7f0ded734"
        ],
        "x": 165,
        "y": 160,
        "wires": [
            [
                "596121eb7d0e536d"
            ]
        ]
    },
    {
        "id": "25038cc00759fcae",
        "type": "function",
        "z": "7bcf78a84f27cb72",
        "name": "dimmerMode=0",
        "func": "flow.set(\"dimmerMode\", 0)\nmsg.reset=true\nreturn msg",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 1220,
        "y": 200,
        "wires": [
            []
        ]
    },
    {
        "id": "1c7df14b7dcc4ddd",
        "type": "function",
        "z": "7bcf78a84f27cb72",
        "name": "normalizeDim",
        "func": "if (msg.payload.action.includes(\"release\")) { msg.payload = \"stop\" }\n\nelse if (msg.payload.action == \"press_1\") {msg.payload = \"up\"}\n\nelse if (msg.payload.action == \"press_2\") {msg.payload = \"down\"}\n\nelse if (msg.payload.action == \"press_3\") {msg.payload = \"up\"}\n\nelse if (msg.payload.action == \"press_4\") {msg.payload = \"down\"}\n\nreturn msg;\n\n",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 560,
        "y": 80,
        "wires": [
            [
                "d00d7c162ab41adc"
            ]
        ]
    },
    {
        "id": "2097d594082c9566",
        "type": "zigbee2mqtt-out",
        "z": "7bcf78a84f27cb72",
        "name": "",
        "server": "e9c6b6542203d706",
        "friendly_name": "BB96E9 (8718696449691)",
        "device_id": "0x0017880104cb5e5b",
        "command": "brightness_step_onoff",
        "commandType": "z2m_cmd",
        "payload": "payload",
        "payloadType": "msg",
        "optionsValue": "",
        "optionsType": "nothing",
        "x": 1290,
        "y": 80,
        "wires": []
    },
    {
        "id": "83529339124975f8",
        "type": "trigger",
        "z": "7bcf78a84f27cb72",
        "name": "",
        "op1": "-25",
        "op2": "10",
        "op1type": "num",
        "op2type": "str",
        "duration": "-500",
        "extend": false,
        "overrideDelay": false,
        "units": "ms",
        "reset": "stop",
        "bytopic": "all",
        "topic": "topic",
        "outputs": 1,
        "x": 1020,
        "y": 100,
        "wires": [
            [
                "2097d594082c9566"
            ]
        ]
    },
    {
        "id": "d00d7c162ab41adc",
        "type": "switch",
        "z": "7bcf78a84f27cb72",
        "name": "",
        "property": "payload",
        "propertyType": "msg",
        "rules": [
            {
                "t": "eq",
                "v": "up",
                "vt": "str"
            },
            {
                "t": "eq",
                "v": "stop",
                "vt": "str"
            },
            {
                "t": "eq",
                "v": "down",
                "vt": "str"
            }
        ],
        "checkall": "true",
        "repair": false,
        "outputs": 3,
        "x": 730,
        "y": 80,
        "wires": [
            [
                "8e67781937d9e961",
                "9d09b86f2050d21b"
            ],
            [
                "8e67781937d9e961",
                "83529339124975f8",
                "9d09b86f2050d21b"
            ],
            [
                "83529339124975f8",
                "9d09b86f2050d21b"
            ]
        ]
    },
    {
        "id": "6ce773ae92d191c0",
        "type": "zigbee2mqtt-out",
        "z": "7bcf78a84f27cb72",
        "name": "",
        "server": "e9c6b6542203d706",
        "friendly_name": "BB96E9 (8718696449691)",
        "device_id": "0x0017880104cb5e5b",
        "command": "brightness",
        "commandType": "z2m_cmd",
        "payload": "payload",
        "payloadType": "msg",
        "optionsValue": "",
        "optionsType": "nothing",
        "x": 1030,
        "y": 280,
        "wires": []
    },
    {
        "id": "e9c6b6542203d706",
        "type": "zigbee2mqtt-server",
        "name": "docker-mqtt-conbee",
        "host": "127.0.0.1",
        "mqtt_port": "1883",
        "mqtt_username": "",
        "mqtt_password": "",
        "mqtt_qos": "0",
        "tls": "",
        "usetls": false,
        "base_topic": "z2m-conbee"
    }
]

**Comment by mbechc** [comment]
Thanks for help and inspiration
